### PR TITLE
Features/1458 add incremental SVD/PCA

### PR DESCRIPTION
**Pull request by mrfh92**
## Due Diligence

- General:
    - [X]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [x] documentation updated where needed

## Description

Issue/s resolved: #1458 
(Should be merged after #1561 as it already contains these changes)

## Changes proposed:

Adds incremental SVD (see M. Brand, Fast low-rank modifications of the thin singular value decomposition, Linear Algebra and its Applications 415 (2006)) and corresponding interface for PCA 

## Type of change
new feature

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by mrfh92** [comment]
@JuanPedroGHM I have merged the rSVD and resolved the conflicts. Now it should be possible to merge after merging rSVD without conflicts.

**Comment by mrfh92** [comment]
@JuanPedroGHM You're right regarding the benchmarks. I will open a separate issue for adding the benchmarks that are not there so far

File: heat/core/linalg/tests/test_svdtools.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import heat as ht
 import numpy as np
 from mpi4py import MPI
+import time
```
Comment (JuanPedroGHM) [inline heat/core/linalg/tests/test_svdtools.py:7]: Probably can be deleted, guessing it was used during development

File: heat/decomposition/tests/test_pca.py [hunk 2]
```diff
@@ -210,3 +209,122 @@ def test_pca_randomized(self):
...
+            ht.decomposition.IncrementalPCA(whiten=True)
+        with self.assertRaises(NotImplementedError):
+            ht.decomposition.IncrementalPCA(copy=False)
+        # wrong n_components
+        with self.assertRaises(TypeError):
+            ht.decomposition.IncrementalPCA(n_components=0.9)
+        with self.assertRaises(ValueError):
+            ht.decomposition.IncrementalPCA(n_components=0)
+
+    def test_incrementalpca_part1(self):
```
Comment (JuanPedroGHM) [inline heat/decomposition/tests/test_pca.py:245]: Use more descriptive names and split test into multiple categories. Part 1 could be `test_incrementalpca_full_rank`, and part 2 could be 'test_incremenatlpca_truncation happens'

**Review by JuanPedroGHM — CHANGES_REQUESTED** [review]
I think iSVD could benefit from a full usage example somewhere, either in the benchmarks or the tests, that we could point people too. The fact that you need to have already done the SVD of a partial matrix, or that `partial_fit` needs to be used, could generate confusion. A simple example of it in use could help a lot in the future.

File: heat/core/linalg/svdtools.py [hunk 3]
```diff
@@ -21,25 +21,24 @@
...
+__all__ = ["hsvd_rank", "hsvd_rtol", "hsvd", "rsvd", "isvd"]
 
 
-def _check_SVD_input(A):
-    if not isinstance(A, DNDarray):
-        raise TypeError(f"Argument needs to be a DNDarray but is {type(A)}.")
-    if not A.ndim == 2:
-        raise ValueError("A needs to be a 2D matrix")
-    if not types.heat_type_is_realfloating(A.dtype):
+def _check_is_nd_realfloating(input, inputname, allowed_ns):
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/linalg/svdtools.py:27]: :+1: 
This is now `ht.sanitation.sanitize_in_nd_realfloating`

File: heat/core/linalg/tests/test_svdtools.py (hunk 1 above)
Comment (mrfh92) [reply heat/core/linalg/tests/test_svdtools.py:7]: done :+1:

File: heat/decomposition/tests/test_pca.py (hunk 2 above)
Comment (mrfh92) [reply heat/decomposition/tests/test_pca.py]: renaming done, introduced another category for catching wrong inputs

**Comment by mrfh92** [comment]
I have added a benchmark for IncrementalPCA that also highlights how this function could be used. I also put a reference in the docs of IncrementalPCA.

**Comment by mrfh92** [comment]
@JuanPedroGHM I have turned off the benchmark_pr as it does not work at the moment, but except of that I guess I have addressed all comments

**Review by JuanPedroGHM — APPROVED** [review]

**Pull request merged**